lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroPageDataReader.java

```diff
@@ -0,0 +1,1065 @@
+import io.trino.spi.block.SingleRowBlockWriter;
+import io.trino.spi.type.RowType;
+import io.trino.spi.type.Type;
+import org.apache.avro.Resolver;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.avro.io.FastReaderBuilder;
+import org.apache.avro.io.parsing.ResolvingGrammarGenerator;
+import org.apache.avro.util.internal.Accessor;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.IntFunction;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.trino.hive.formats.avro.AvroTypeUtils.UNION_FIELD_TAG_TYPE;
+import static io.trino.hive.formats.avro.AvroTypeUtils.isSimpleNullableUnion;
+import static io.trino.hive.formats.avro.AvroTypeUtils.typeFromAvro;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.lang.Float.floatToRawIntBits;
+import static java.util.Objects.requireNonNull;
+
+public class AvroPageDataReader
+        implements DatumReader<Optional<Page>>
+{
+    private final Schema readerSchema;
+    private Schema writerSchema;
+    private final PageBuilder pageBuilder;
+    private RowBlockBuildingDecoder rowBlockBuildingDecoder;
+    private final AvroTypeManager typeManager;
+
+    public AvroPageDataReader(Schema readerSchema, AvroTypeManager typeManager)
+            throws AvroTypeException
+    {
+        try {
+            this.readerSchema = requireNonNull(readerSchema, "readerSchema is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+            this.writerSchema = this.readerSchema;
+            Type readerSchemaType = typeFromAvro(this.readerSchema, typeManager);
+            verify(readerSchemaType instanceof RowType, "Can only build pages when top level type is Row");
+            this.pageBuilder = new PageBuilder(readerSchemaType.getTypeParameters());
+            initialize();
+        }
+        catch (org.apache.avro.AvroTypeException runtimeAvroException) {
+            throw new AvroTypeException(runtimeAvroException.getMessage(), runtimeAvroException.getCause());
+        }
```

Comment:
I'd rewrite this to:
* move simple field initialization outside of try/catch so it is easy to see what code might throw the exception
* I find the error message about the root type to be a bit confusing... I suggested something different, but mine isn't great either
* for the exception handling, I don't unwrap the cause as that can lose context that make it hard to debug
```
        this.readerSchema = requireNonNull(readerSchema, "readerSchema is null");
        this.writerSchema = this.readerSchema;
        this.typeManager = requireNonNull(typeManager, "typeManager is null");
        try {
            Type readerSchemaType = typeFromAvro(this.readerSchema, typeManager);
            verify(readerSchemaType instanceof RowType, "Root Avro type must be a row");
            this.pageBuilder = new PageBuilder(readerSchemaType.getTypeParameters());
            initialize();
        }
        catch (org.apache.avro.AvroTypeException avroTypeException) {
            throw new AvroTypeException(avroTypeException.getMessage(), avroTypeException);
        }
```